pkg/cluster/kubeconfig.go

```diff
@@ -18,26 +18,70 @@ import (
 
 	"github.com/Azure/ARO-RP/pkg/api"
 	"github.com/Azure/ARO-RP/pkg/cluster/graph"
+	utilpem "github.com/Azure/ARO-RP/pkg/util/pem"
 	"github.com/Azure/ARO-RP/pkg/util/stringutils"
 )
 
 // generateAROServiceKubeconfig generates additional admin credentials and a
 // kubeconfig for the ARO service, based on the admin kubeconfig found in the
 // graph.
 func (m *manager) generateAROServiceKubeconfig(pg graph.PersistedGraph) (*kubeconfig.AdminInternalClient, error) {
-	return generateKubeconfig(pg, "system:aro-service", []string{"system:masters"}, tls.ValidityTenYears)
+	return generateKubeconfig(pg, "system:aro-service", []string{"system:masters"}, tls.ValidityTenYears, true)
 }
 
 // generateAROSREKubeconfig generates additional admin credentials and a
 // kubeconfig for ARO SREs, based on the admin kubeconfig found in the graph.
 func (m *manager) generateAROSREKubeconfig(pg graph.PersistedGraph) (*kubeconfig.AdminInternalClient, error) {
-	return generateKubeconfig(pg, "system:aro-sre", nil, tls.ValidityTenYears)
+	return generateKubeconfig(pg, "system:aro-sre", nil, tls.ValidityTenYears, true)
+}
+
+// checkUserAdminKubeconfigUpdated checks if the user kubeconfig is
+// present, has >90days until expiry, has the right settings
+func (m *manager) checkUserAdminKubeconfigUpdated(ctx context.Context) bool {
+	if len(m.doc.OpenShiftCluster.Properties.UserAdminKubeconfig) == 0 {
+		// field empty, not updated
+		return false
+	}
+	var aic kubeconfig.AdminInternalClient
+	err := yaml.Unmarshal([]byte(m.doc.OpenShiftCluster.Properties.UserAdminKubeconfig), &aic.Config)
+	if err != nil {
+		// yaml invalid, not updated
+		return false
+	}
+	for i := range aic.Config.Clusters {
+		if strings.HasPrefix(aic.Config.Clusters[i].Cluster.Server, "https://api-int.") {
+			// URL pointing to api-int, not updated
+			// TODO remove this after PUCM has been run on all clusters.
+			return false
+		}
+	}
+	for i := range aic.Config.AuthInfos {
+		innerpem := string(aic.Config.AuthInfos[i].AuthInfo.ClientCertificateData) + string(aic.Config.AuthInfos[i].AuthInfo.ClientKeyData)
+		innerkey, innercert, err := utilpem.Parse([]byte(innerpem))
+		if err != nil {
+			// error while parsing cert or key, not updated
+			return false
+		}
+		if innerkey == nil {
+			// no client key, not updated
+			return false
+		}
+		for j := range innercert {
+			if !innercert[j].NotAfter.After(time.Now().AddDate(0, 0, 90)) {
```

Comment:
I had to write a simplified version to understand this. I think it is very easy to break and we need a good test coverage for this bit. I think we need to turn `TestGenerateUserAdminKubeconfig` into a test with multiple test cases. Or add a separate unit test to test `checkUserAdminKubeconfigUpdated`.



```go
package main

import (
	"fmt"
	"time"
)

func main() {
	notAfter := time.Date(2022, time.January, 1, 0, 0, 0, 0, time.UTC)
	fmt.Println("notAfter      ", notAfter)
	fmt.Println("============================================")

	for _, now := range []time.Time{
		// now == (notAfter-91d) - expires in more than 90 days - willUpdate == false
		notAfter.AddDate(0, 0, -91),
		// now == notAfter - expires right now - willUpdate == true
		notAfter,
		// now == (notAfter-90d) - expires exactly in 90 days - willUpdate == true
		notAfter.AddDate(0, 0, -90),
		// now == (notAfter+90d) - already expired - willUpdate == true
		notAfter.AddDate(0, 0, 90),
	} {
		fmt.Println("willUpdate    ", !notAfter.After(now.AddDate(0, 0, 90)))
		fmt.Println("--------------------------------------------")
	}
}

```

---

I checked and there is no coverage for `checkUserAdminKubeconfigUpdated`. We need to fix it.

![image](https://user-images.githubusercontent.com/509198/139912176-bbf3a279-821e-43a9-8e3e-c7c39cd5f22b.png)

Reply:
Added new unit test for this func.